cni/pkg/nodeagent/ztunnelserver.go

```diff
@@ -0,0 +1,480 @@
+					Payload: &zdsapi.WorkloadRequest_Keep{
+						Keep: &zdsapi.KeepWorkload{
+							Uid: uid,
+						},
+					},
+				}, nil)
+			}
+			if err != nil {
+				return err
+			}
+			if resp.GetAck().GetError() != "" {
+				log.Errorf("add-workload: got ack error: %s", resp.GetAck().GetError())
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	resp, err := conn.sendMsgAndWaitForAck(&zdsapi.WorkloadRequest{
+		Payload: &zdsapi.WorkloadRequest_SnapshotSent{
+			SnapshotSent: &zdsapi.SnapshotSent{},
+		},
+	}, nil)
+	if err != nil {
+		return err
+	}
+	log.Debugf("snaptshot sent to ztunnel")
+	if resp.GetAck().GetError() != "" {
+		log.Errorf("snap-sent: got ack error: %s", resp.GetAck().GetError())
+	}
+
+	return nil
+}
+
+func (z *ztunnelServer) accept() (*ZtunnelConnection, error) {
+	log.Debug("accepting unix conn")
+	conn, err := z.listener.AcceptUnix()
+	if err != nil {
+		return nil, fmt.Errorf("failed to accept unix: %w", err)
+	}
+	log.Debug("accepted conn")
+	return newZtunnelConnection(conn), nil
+}
+
+type updateResponse struct {
+	err  error
+	resp *zdsapi.WorkloadResponse
+}
+
+type updateRequest struct {
+	Update []byte
+	Fd     *int
+
+	Resp chan updateResponse
+}
+
+type ZtunnelConnection struct {
+	u       *net.UnixConn
+	Updates chan updateRequest
+}
+
+func newZtunnelConnection(u *net.UnixConn) *ZtunnelConnection {
+	return &ZtunnelConnection{u: u, Updates: make(chan updateRequest, 100)}
+}
+
+func (z *ZtunnelConnection) Close() {
+	z.u.Close()
+}
+
+func (z *ZtunnelConnection) send(ctx context.Context, data []byte, fd *int) (*zdsapi.WorkloadResponse, error) {
+	ret := make(chan updateResponse, 1)
+	req := updateRequest{
+		Update: data,
+		Fd:     fd,
+		Resp:   ret,
+	}
+	select {
+	case z.Updates <- req:
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+	select {
+	case r := <-ret:
+		return r.resp, r.err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (z *ZtunnelConnection) sendMsgAndWaitForAck(msg *zdsapi.WorkloadRequest, fd *int) (*zdsapi.WorkloadResponse, error) {
+	data, err := proto.Marshal(msg)
+	if err != nil {
+		return nil, err
+	}
+	return z.sendDataAndWaitForAck(data, fd)
+}
+
+func (z *ZtunnelConnection) sendDataAndWaitForAck(data []byte, fd *int) (*zdsapi.WorkloadResponse, error) {
+	var rights []byte
+	if fd != nil {
+		rights = unix.UnixRights(*fd)
+	}
+	err := z.u.SetWriteDeadline(time.Now().Add(readWriteDeadline))
+	if err != nil {
+		return nil, err
+	}
+
+	_, _, err = z.u.WriteMsgUnix(data, rights, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// wait for ack
+	return z.readMessage(readWriteDeadline)
+}
+
+func (z *ZtunnelConnection) readMessage(timeout time.Duration) (*zdsapi.WorkloadResponse, error) {
+	m, _, err := readProto[zdsapi.WorkloadResponse](z.u, timeout, nil)
+	return m, err
+}
+
+func readProto[T any, PT interface {
+	proto.Message
+	*T
+}](c *net.UnixConn, timeout time.Duration, oob []byte) (PT, int, error) {
```

Comment:
Nit: can just have one param and make them pass `*zdsapi.WorkloadResponse`. Any usage of `T` is invalid anyways

Reply:
We reuse this quite a bit in tests as a generic `readProto` func for stub ztunnelClients reading `zdsapi.WorkloadRequest`'s as well.

We could dupe the code but it's the same code for both on either side of the connection, so it makes sense to keep it in one spot.